File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/ArrayOf1AsCollectionHolder.java
package joookiwi.collection.java;

import java.lang.ref.WeakReference;
import java.util.function.Consumer;
import java.util.function.IntConsumer;
import java.util.function.ObjIntConsumer;
import joookiwi.collection.java.callback.IntObjConsumer;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.MustBeInvokedByOverriders;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_NEW_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;

/// An instance of [CollectionHolder] adapted from an `array` having a lone value inside.
///
/// Note that the value is directly retrieved from the array and it is kept
///
/// @param <T> The type
/// @see ArrayAsCollectionHolder
/// @see CollectionHolderOf1
/// @see LazyCollectionHolderOf1
@NotNullByDefault
public class ArrayOf1AsCollectionHolder<T extends @Nullable Object>
        extends AbstractCollectionHolderOf1<T> {

    private final WeakReference<T @Unmodifiable []> __reference;
    private final T __value;
    private final boolean __hasNull;
    private final boolean __hasNoNulls;

    public ArrayOf1AsCollectionHolder(final T @Unmodifiable [] reference) {
        super();
        if (reference.length != 1)
            throw new IllegalArgumentException("The array received in the " + getClass().getSimpleName() + " cannot have a different size than 1.");
        __reference = new WeakReference<>(reference);
        __hasNoNulls = !(__hasNull = (__value = reference[0]) == null);
    }

    /// The internal value passed through the `constructor` in the [#_reference] first field
    @Override public T value() { return __value; }

    /// The internal referenced passed through the `constructor`.
    ///
    /// It returns `null` if the reference has been garbage-collected by the **JVM**
    protected T @Nullable @Unmodifiable [] _reference() { return __reference.get(); }

    @Override public boolean hasNull() { return __hasNull; }

    @Override public boolean hasNoNulls() { return __hasNoNulls; }

    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEach(final ObjIntConsumer<? super T> action) { return (ArrayOf1AsCollectionHolder<T>) super.onEach(action); }
    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEach(final Consumer<? super T>       action) { return (ArrayOf1AsCollectionHolder<T>) super.onEach(action); }
    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEach(final Runnable                  action) { return (ArrayOf1AsCollectionHolder<T>) super.onEach(action); }

    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEachIndexed(final IntObjConsumer<? super T> action) { return (ArrayOf1AsCollectionHolder<T>) super.onEachIndexed(action); }
    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEachIndexed(final IntConsumer               action) { return (ArrayOf1AsCollectionHolder<T>) super.onEachIndexed(action); }
    @Contract(ALWAYS_THIS_1) @Override public ArrayOf1AsCollectionHolder<T> onEachIndexed(final Runnable                  action) { return (ArrayOf1AsCollectionHolder<T>) super.onEachIndexed(action); }



    @MustBeInvokedByOverriders
    @Contract(ALWAYS_NEW_0)
    @Override public ArrayOf1AsCollectionHolder<T> clone() { return (ArrayOf1AsCollectionHolder<T>) super.clone(); }

//    @MustBeInvokedByOverriders
//    @Contract(ALWAYS_NEW_0)
//    @Override public ArrayOf1AsCollectionHolder<T> shallowClone() { return clone(); }
//
//    @MustBeInvokedByOverriders
//    @Contract(ALWAYS_NEW_0)
//    @Override public ArrayOf1AsCollectionHolder<T> deepClone() { return new ArrayOf1AsCollectionHolder<>(CloneInvokator.getInstance().clone(_reference())); }

}

File: javascript/src/ArrayOf1AsCollectionHolder.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {UndefinedOr} from "@joookiwi/type"

import {AbstractCollectionHolderOf1} from "./AbstractCollectionHolderOf1"

/**
 * An instance of {@link CollectionHolder} adapted from an {@link ReadonlyArray Array} having a lone value inside.
 *
 * Note that the value is directly retrieved from the array and it is kept
 *
 * @typeParam T The type
 * @see ArrayAsCollectionHolder
 * @see SetOf1AsCollectionHolder
 * @see CollectionHolderOf1
 * @see LazyCollectionHolderOf1
 */
export class ArrayOf1AsCollectionHolder<const T = unknown,
    const REFERENCE extends readonly [T,] = readonly [T,], >
    extends AbstractCollectionHolderOf1<T> {

    /** The internal value passed through the {@link constructor} in the {@link _reference} first field */
    public override readonly 0: T
    readonly #reference: WeakRef<REFERENCE>
    readonly #value: T
    readonly #hasNull: boolean
    readonly #hasNoNulls: boolean

    public constructor(reference: REFERENCE,) {
        super()
        if (reference.length !== 1)
            throw new TypeError(`The array received in the “${this.constructor.name}” cannot have a different size than 1.`,)
        this.#reference = new WeakRef(reference,)
        this.#hasNoNulls = !(this.#hasNull = (this.#value = this[0] = reference[0]) == null)
    }

    /** The internal value passed through the {@link constructor} in the {@link _reference} first field */
    public override get value(): T { return this.#value }

    /**
     * The internal referenced passed through the {@link constructor}.
     *
     * It returns `undefined` if the reference has been garbed-collected.
     */
    protected get _reference(): UndefinedOr<REFERENCE> { return this.#reference.deref() }

    public override get hasNull(): boolean { return this.#hasNull }

    public override get hasNoNulls(): boolean { return this.#hasNoNulls }

}
